### ISSUE TITLE
Rule Request: Prefer returning `Self` in value types instead of the value type's name

### ISSUE THREAD

**Issue by Noobish1** [realm/SwiftLint#2944]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

This rule is similar to https://github.com/realm/SwiftLint/issues/2921 but I don't think it's a duplicate.

### New rule request

Please describe the rule idea, format
this issue's title as `Rule Request: [Rule Name]` and describe:

1. Why should this rule be added?

I can't find the exact Swift proposal but in Swift 5.1 (I think) you can return `Self` in value types instead of the value type's name.

e.g.

```swift
struct Foo {
    static func someFunc() -> Self { ... }
}
```

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

would trigger:
```swift
struct Foo {
    static func someFunc() -> Foo { ... }
}
```

```swift
struct Foo {
    func someFunc() -> Foo { ... }
}
```

```swift
struct Foo {
    var someComputedProperty: Foo { ... }
}
```

wouldn't trigger:
```swift
struct Foo {
    func someFunc() -> Self { ... }
}
```

```swift
class Foo {
    func someFunc() -> Foo { ... }
}
```

3. Should the rule be configurable, if so what parameters should be configurable?

I don't think so

4. Should the rule be opt-in or enabled by default? Why?
   See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a rule as opt-in.

Safer to opt-in I guess.
